### ISSUE TITLE
Prisma create failing with query interpretation errors

### ISSUE THREAD

**Issue by nilaeus** [prisma/prisma#2087]
## Bug description

When trying to create a new record (with many other actions nested within), schema works fine, everything looks good with TypeScript, but it fails when run.

[Debug output](https://jpst.it/25Ack)

## How to reproduce

(With the files below, prisma information section)
1. `npm i -g ts-node`
2. `npm i @prisma/cli @prisma/client`
3. `prisma migrate save --experimental && prisma migrate up --experimental && prisma generate`
4. `ts-node error.ts`

## Expected behavior

Prisma should create a `Person` record, with all the data, create and link a `Block` record to `personBlock`, that is owned by the same `Person` record. The `Block` should also have a couple `internalRelations`, which each create a `Block` and connect to a `Person` record, in this case the same one.

## Prisma information

I'll try my best to answer questions or try things out quickly if needed.

### schema.prisma
Disclaimer, the schema file contains many items that are not touched on in the code that errors.
```prisma
datasource db {
  provider = "postgresql"
  url      = "postgresql://prisma:prisma@localhost:5432/loop"
}

generator photon {
  provider = "prisma-client-js"
}

model Person {
  // Basic info
  id            Int         @default(autoincrement()) @id
  name          String
  checkingName  String      @unique
  password      String
  joined        DateTime    @default(now())
  // Credits
  credits       Int         @default(100)
  streak        Int         @default(2)
  lastCheckedIn DateTime?
  // Blocks
  ownedBlocks   Block[]     @relation("ownership", references: [id])
  personBlock   Block       @relation(references: [id])
  hasAccessTo   Block[]     @relation("access", references: [id])
  typesOwned    BlockType[]
  setings       Settings[]  @relation("settingsPerson", references: [id])
}

model Block {
  id                Int                @default(autoincrement()) @id
  type              BlockType
  data              String?
  // Permission
  owner             Person             @relation("ownership", fields: [ownerId], references: [id])
  peopleWithAccess  Person[]           @relation("access", references: [id])
  // Other blocks in relation to this one
  internalRelations InternalRelation[] @relation("internalOf", references: [id])
  externalBlocks    Block[]            @relation("external", references: [id])
  personalSetttings Settings[]         @relation("settingsBlock", references: [id])
  // This block in relation to other blocks
  isInternalOf      InternalRelation[] @relation("propertyValue", references: [id])
  isExternalOf      Block[]            @relation("external", references: [id])
  isPersonBlockOf   Person?
  // Is setting of
  isDisplayNameOf   Settings[]         @relation("displayName", references: [id])
  // Relation junk
  ownerId           Int
}

model InternalRelation {
  id              Int     @default(autoincrement()) @id
  propertyName    String?
  propertyValue   Block   @relation("propertyValue", fields: [propertyValueId], references: [id])
  isPropertyOf    Block   @relation("internalOf", fields: [isPropertyOfId], references: [id])
  // Relation junk
  propertyValueId Int
  isPropertyOfId  Int
}

model Settings {
  id                   Int    @default(autoincrement()) @id
  isSettingsOfBlock    Block  @relation("settingsBlock", fields: [isSettingsOfBlockId], references: [id])
  isSettingsOfPerson   Person @relation("settingsPerson", fields: [isSettingsOfPersonId], references: [id])
  // Settings
  displayName          Block  @relation("displayName", fields: [isDisplayNameOf], references: [id])
  // Relation junk
  isSettingsOfBlockId  Int
  isSettingsOfPersonId Int
  isDisplayNameOf      Int
}

enum BlockType {
  Group
  Number
  Person
  RawText
  Text
}
```

### error.ts
```typescript
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

const prettyMuch = async () => {
  const encrypted = 'substitute'
  const name = 'Loop'
  const checkingName = 'loop'
  const connectLoop = { connect: { checkingName: 'loop' } }
  const person = await prisma.person.create({
    data: {
      checkingName,
      name,
      password: encrypted,
      personBlock: {
        create: {
          data: checkingName,
          internalRelations: {
            create: [
              {
                propertyName: 'name',
                propertyValue: {
                  create: {
                    data: name,
                    owner: connectLoop,
                    type: 'RawText',
                  },
                },
              },
              {
                propertyName: 'created',
                propertyValue: {
                  create: {
                    owner: connectLoop,
                    type: 'Group',
                  },
                },
              },
              {
                propertyName: 'credits',
                propertyValue: {
                  create: {
                    data: '100 credits',
                    owner: connectLoop,
                    type: 'Number',
                  },
                },
              },
              {
                propertyName: 'streak',
                propertyValue: {
                  create: {
                    data: '2 days',
                    owner: connectLoop,
                    type: 'Number',
                  },
                },
              },
            ],
          },
          owner: {
            connect: { checkingName },
          },
          type: 'Person',
        },
      },
      typesOwned: { set: ['Group', 'Number', 'Text', 'RawText'] },
    },
  })
}

prettyMuch()
  .then(console.log)
  .catch(console.error)
```

## Environment & setup

I'm on MacOS Catalina 10.15.4
Using Postgresql provided by Prisma
Prisma version says `2.0.0-beta.1` and darwin
Node `12.14.0`
I run it with `ts-node`

[Again, here is the error](https://jpst.it/25Ack)

**Comment by ranjan-purbey** [comment]
Can you also share error logs?

**Comment by ranjan-purbey** [comment]
The `owner connect` object should have a `where` key, I think

**Comment by nilaeus** [comment]
@ranjan-purbey Connecting to the owner record does not need a `where` key, or at least TypeScript will error. I believe that this is a Prisma issue to an extent because the compile-time checks out, but runtime does not, and the error is illegible to me.

The error logs are pretty lengthy, so I put them on [this link](https://jpst.it/25Ack).

**Comment by ranjan-purbey** [comment]
My bad, I got confused. So two changes needed here:
1. The property `personBlock` on `Person` model should have a corresponding scalar field. This scalar field is used to store the foreign key in database table.
2. The `Person` model has non-nullable field `personBlock` which means while creating a person, you must either create or connect a `Block`. Also the `owner` field on `Block` is mandatory, so when creating a `Block` you also must create/connect a `User`. To fix this, you need to make one of them nullable

**Comment by nilaeus** [comment]
Thank you so much for helping.
Adding the relations as I'd expect:
```prisma
// In addition to what's already there
model Person {
    personBlock   Block  @relation("personBlock", fields: [personBlockId], references: [id])
    personBlockId Int
}
model Block {
    isPersonBlockOf Person? @relation("personBlock", references: [id])
}
```
Errors with
```batch
 ERROR  Oops, an unexpected error occured!
Error in migration engine.
Reason: [libs/prisma-models/src/datamodel_converter.rs:261:33] It's not allowed that both sides of a relation
 specify the inline policy. The field was personBlock on model Person. The related field was isPersonBlockOf 
on model Block.
```
I found that
```prisma
// In addition to what's already there
model Person {
    personBlock   Block  @relation(fields: [personBlockId], references: [id])
    personBlockId Int
}
model Block {
    isPersonBlockOf Person?
}
```
(Removing the ®relation name and explicit relation on `isPersonBlockOf`)
Works as far as schema goes.

To your second point, is that an issue? I remember that working in Prisma 1, and I think that it should still work. `isPersonOf` is optional on a `Block`, and this first execution of the code that creates the first `Person` should have ownership of these blocks.

I tested all combinations of `personBlock` and `owner` being nullable, but it still throws the same error.
When it is optional, removing all the `owner`s when creating `Block`s does work, but I do need that relation. In case it was becuase the person `checkingName` or `connectLoop` (both loop) didn't exist, I first made a person with the checkingName of `loop`, then tried it again with a different name. I errored the same way.

I think that there is definitely something fishy with `owner` relations, but I do not know what.

Thanks again for helping

**Comment by ranjan-purbey** [comment]
I don't think such operation is allowed. What you can do is make the personBlock field optional. Create the person with the blocks and then update the personBlock of the newly created person in a separate call.

**Comment by nilaeus** [comment]
I've updated `error.ts` to:
```typescript
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

const prettyMuch = async () => {
  const encrypted = 'substitute'
  const name = 'Bill'
  const checkingName = 'bill'
  const connectLoop = { connect: { checkingName: 'loop' } }
  const person = await prisma.person.create({
    data: {
      checkingName,
      name,
      password: encrypted,
      typesOwned: { set: ['Group', 'Number', 'Text', 'RawText'] },
    },
  })
  await prisma.block.create({
    data: {
      data: checkingName,
      internalRelations: {
        create: [
          {
            propertyName: 'name',
            propertyValue: {
              create: {
                data: name,
                owner: connectLoop,
                type: 'RawText',
              },
            },
          },
          {
            propertyName: 'created',
            propertyValue: {
              create: {
                owner: connectLoop,
                type: 'Group',
              },
            },
          },
          {
            propertyName: 'credits',
            propertyValue: {
              create: {
                data: '100 credits',
                owner: connectLoop,
                type: 'Number',
              },
            },
          },
          {
            propertyName: 'streak',
            propertyValue: {
              create: {
                data: '2 days',
                owner: connectLoop,
                type: 'Number',
              },
            },
          },
        ],
      },
      isPersonBlockOf: { connect: { checkingName } },
      owner: {
        connect: { checkingName },
      },
      type: 'Person',
    },
  })
}

prettyMuch()
  .then(console.log)
  .catch(console.error)
```
And now [another issue](https://jpst.it/25EwM) occurs.
As expected, it now creates a Person, but the Block errs.

**Comment by nilaeus** [comment]
I made `owner` optional again, and commented out where it calls to connect an owner, and it works. Connecting with `isPersonBlockOf` does work, however. I think there is something with the `owner`, so maybe this is more of an issue with my schema

**Comment by pantharshit00** [comment]
@nilaeus Sorry for the late reply here.

Can you tell me what problem you have now after that triaging you did with @ranjan-purbey?

It it hard to follow after all of these comments. So can you give me a brief summery now?

**Comment by nilaeus** [comment]
I still feel like the issue I first opened this for shouldn't happen, and to not complicate things I'd like your final input on that before talking about the new issue.

With my first code, it all looks fine (to me and the compiling), but then something errors:
```fish
Error occurred during query execution:
InterpretationError("Error for binding \'14\': AssertionError(\"[Query Graph] Expected a valid parent ID to be present for a nested connect on a one-to-many relation.\")")
```
[Here is the full debug output, just in case](https://jpst.it/25Ack)

The reason I still would like confirmation on this is that any way about it there is an issue because the logs were too vague.

I found that making the `owner` field on the `block` model optional, then commenting out places where an owner is connected, works without errors. I think that there is something going on with the `owner` field.

Oops I guess this isn't that brief, but thank you for helping, @pantharshit00

**Comment by nilaeus** [comment]
I've been working with a workaround of changing `owner` to be the name of the owner person. It's been working so far in terms of avoiding this problem, but I've encountered another error that I think may be of the same vein.
It turns out that the above `prisma.person.create()` does not error when the owner is just a string, but it **does not create** the `InternalRelation`s. It creates the blocks inside, but not the InternalRelations themselves, which is very odd.

Then, I tried to make InternalRelations on their own, and ran into this error:
```
Error in Prisma Client:
PANIC: assertion failed: index < len in
<::core::macros::panic macros>:2:2

This is a non-recoverable error which probably happens when the Prisma Query Engine has a panic.
```

This seems a little similar to [this issue](https://github.com/prisma/prisma-client-js/issues/633) in prisma-client-js, should this issue be moved there @pantharshit00 ?

**Comment by pantharshit00** [comment]
Sorry for the delay. I can confirm this issue. Looks like an issue on our end. Your datamodel is bit out of data though. Here is the datamodel which is valid:
```prisma
model Person {
  // Basic info
  id            Int         @default(autoincrement()) @id
  name          String
  checkingName  String      @unique
  password      String
  joined        DateTime    @default(now())
  // Credits
  credits       Int         @default(100)
  streak        Int         @default(2)
  lastCheckedIn DateTime?
  // Blocks
  ownedBlocks   Block[]     @relation("ownership")
  personBlock   Block       @relation(references: [id], fields:[personBlockId])
  personBlockId Int
  hasAccessTo   Block[]     @relation("access", references: [id])
  typesOwned    BlockType[]
  setings       Settings[]  @relation("settingsPerson")
}

model Block {
  id                Int                @default(autoincrement()) @id
  type              BlockType
  data              String?
  // Permission
  owner             Person @relation("ownership", fields: [ownerId], references: [id])
  peopleWithAccess  Person[]           @relation("access", references: [id])
  // Other blocks in relation to this one
  internalRelations InternalRelation[] @relation("internalOf")
  externalBlocks    Block[]            @relation("external", references: [id])
  personalSetttings Settings[]         @relation("settingsBlock")
  // This block in relation to other blocks
  isInternalOf      InternalRelation[] @relation("propertyValue")
  isExternalOf      Block[]            @relation("external", references: [id])
  isPersonBlockOf   Person?
  // Is setting of
  isDisplayNameOf   Settings[]         @relation("displayName")
  // Relation junk
  ownerId           Int
}

model InternalRelation {
  id              Int     @default(autoincrement()) @id
  propertyName    String?
  propertyValue   Block   @relation("propertyValue", fields: [propertyValueId], references: [id])
  isPropertyOf    Block   @relation("internalOf", fields: [isPropertyOfId], references: [id])
  // Relation junk
  propertyValueId Int
  isPropertyOfId  Int
}

model Settings {
  id                   Int    @default(autoincrement()) @id
  isSettingsOfBlock    Block  @relation("settingsBlock", fields: [isSettingsOfBlockId], references: [id])
  isSettingsOfPerson   Person @relation("settingsPerson", fields: [isSettingsOfPersonId], references: [id])
  // Settings
  displayName          Block  @relation("displayName", fields: [isDisplayNameOf], references: [id])
  // Relation junk
  isSettingsOfBlockId  Int
  isSettingsOfPersonId Int
  isDisplayNameOf      Int
}

enum BlockType {
  Group
  Number
  Person
  RawText
  Text
}

```

The script that can be used for reproduction:
```js
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

const prettyMuch = async () => {
  const encrypted = "substitute";
  const name = "Loop";
  const checkingName = "loop";
  const connectLoop = { connect: { checkingName: "loop" } };
  const person = await prisma.person.create({
    data: {
      checkingName,
      name,
      password: encrypted,
      personBlock: {
        create: {
          data: checkingName,
          internalRelations: {
            create: [
              {
                propertyName: "name",
                propertyValue: {
                  create: {
                    data: name,
                    owner: connectLoop,
                    type: "RawText",
                  },
                },
              },
              {
                propertyName: "created",
                propertyValue: {
                  create: {
                    owner: connectLoop,
                    type: "Group",
                  },
                },
              },
              {
                propertyName: "credits",
                propertyValue: {
                  create: {
                    data: "100 credits",
                    owner: connectLoop,
                    type: "Number",
                  },
                },
              },
              {
                propertyName: "streak",
                propertyValue: {
                  create: {
                    data: "2 days",
                    owner: connectLoop,
                    type: "Number",
                  },
                },
              },
            ],
          },
          owner: {
            connect: { checkingName },
          },
          type: "Person",
        },
      },
      typesOwned: { set: ["Group", "Number", "Text", "RawText"] },
    },
  });
  console.log(person);
};

prettyMuch().then(console.log).catch(console.error);
```

Error thrown:
```
PrismaClientKnownRequestError:
Invalid `prisma.person.create()` invocation in
/Users/harshit/code/reproductions/issue-2087-2/index.ts:10:38

   6 const encrypted = "substitute";
   7 const name = "Loop";
   8 const checkingName = "loop";
   9 const connectLoop = { connect: { checkingName: "loop" } };
→ 10 const person = await prisma.person.create(

Query interpretation error. Error for binding '14': AssertionError("[Query Graph] Expected a valid parent ID to be present for a nested connect on a one-to-many relation.")
    at PrismaClientFetcher.originalMethod [as request] (/Users/harshit/code/reproductions/issue-2087-2/node_modules/@prisma/client/src/runtime/getPrismaClient.ts:639:11)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  code: 'P2016',
  meta: {
    details: `Error for binding '14': AssertionError("[Query Graph] Expected a valid parent ID to be present for a nested connect on a one-to-many relation.")`
  }
}
```

**Comment by dpetrick** [comment]
Potentially fixed on alpha-1141 onwards. Please check if the issue is resolved.

**Comment by nilaeus** [comment]
Updated to `alpha.1142`, issue is still there:
```
Query interpretation error. Error for binding '14': AssertionError("[Query Graph] Expected a valid parent ID to be present for a nested connect on a one-to-many relation.")
    at PrismaClientFetcher.isPanic (/Users/nilaeus/Desktop/GitHub/blocks/node_modules/@prisma/client/src/runtime/getPrismaClient.ts:633:19)
    at processTicksAndRejections (internal/process/task_queues.js:93:5) {
  code: 'P2016',
  meta: {
    details: `Error for binding '14': AssertionError("[Query Graph] Expected a valid parent ID to be present for a nested connect on a one-to-many relation.")`
  }
}
```

**Comment by dpetrick** [comment]
I looked closer at your schema and the initial operation (with owner required etc.) you are trying to do is not possible and should be prohibited on our side. Prisma needs to order your operations so that all relations and their inlining (e.g. on which table the relation ID is stored) can be satisfied.
The operation you're doing basically does the following:
- You want to create `Person`, and the first nested operation you want to do is to create a `Block`. The ID of that relation is on `Person`, and the relation is required, meaning that we can't create `Person` first, or else we assume that the underlying database throws invalid (simplified, e.g. because the field is marked required in the schema, we assume that it's marked as `NOT NULL`).
- So we need to create `Block` first to get the necessary data to create `Person` in a follow-up. However, the required relation to `owner` states that the inlined ID for that relation is on `Block`, so in order to create the nested record... we need the parent record created, which is impossible to satisfy.

Furthermore, concerning the actual error you see: You're trying to connect a record that is in the process of being created. even if the above would be possible in any way, there's a good chance chance that `Person` with `checkedName` doesn't exist yet because of the ordering or operations that I described. We can't disallow this on a type level really, but a good rule to stick to is that you can't assume that the data you create is also immediately available further down the query tree just because it's required by a dependent query.

Tl;dr - There is no ordering of queries possible that makes the create possible.

**Comment by nilaeus** [comment]
Oh, I see.
Sorry this turned out to be more of a personal issue, when I have time I'll fix my code.
Is there still an issue here, then?

**Comment by dpetrick** [comment]
I don't believe there is, the underlying issue - disallowing queries like that, is something we're aware of but can't fully articulate right now, as the issue is quite complicated. In time I will open a separate issue that encompasses similar issues, but for now I will close this one.

And no worries, thanks for reporting, it's appreciated!
